R/utils.R

```diff
@@ -81,3 +56,63 @@ make_c_call <- function(choices) {
   if (length(private$session_bindings) > 0) lapply(private$session_bindings, function(x) x$destroy())
   invisible(NULL)
 }
+
+
+
+#' Encodes ids to be used in JavaScript and Shiny
+#'
+#' @description
+#' Replaces non-ASCII characters into a format that can be used in HTML,
+#' JavaScript and Shiny.
+#'
+#' When the id has a character that is not allowed, it is replaced with `"_"`
+#' and a hash of the original id is added to the beginning of the id.
+#'
+#' @param id (`character(1)`) The id string.
+#'
+#' @return Sanitized string that removes special characters and spaces.
+#'
+#' @keywords internal
+sanitize_id <- function(id) {
+  escape_characters <- " !\"#$%&'()*+,./:;<=>?@[\\]^`{|}~"
+  pattern <- paste(
+    sep = "", collapse = "|", "\\", strsplit(escape_characters, "")[[1]]
+  )
+  pattern <- gsub("\\<", "<", pattern, fixed = TRUE)
+  pattern <- gsub("\\>", ">", pattern, fixed = TRUE)
+
+  id_new <- gsub(pattern, "_", id)
+  hashes <- vapply(
+    id[id != id_new],
+    rlang::hash, character(1),
+    USE.NAMES = FALSE
+  )
+
+  id[id != id_new] <- paste0("h", substr(hashes, 1, 4), "_", id_new[id != id_new])
+  id
+}
+
+#' `NS` wrapper to sanitize ids for shiny
+#'
+#' Special characters and spaces are not allowed in shiny ids (in JS)
+#'
+#' @noRd
+NS <- function(namespace, id = NULL) { # nolint: object_name.
+  if (!missing(id)) {
+    return(shiny::NS(namespace, sanitize_id(id)))
+  }
+
+  function(id) {
+    shiny::NS(namespace, sanitize_id(id))
+  }
+}
+
+#' `moduleServer` wrapper to sanitize ids for shiny
+#'
+#' Special characters and spaces are not allowed in shiny ids (in JS)
+#'
+#' @noRd
+moduleServer <- function(id, module, session = getDefaultReactiveDomain()) { # nolint: object_name.
+  id <- sanitize_id(id)
+  shiny::moduleServer(id, module, session)
```

Comment:
There is a potential problem with this PR, as a call to `session$ns(" a%problem ")` inside a moduleServer doesn't sanitize the id.

We can:

1. Live with this limitation (as it doesn't appear to affect us)
2. Modify the environment of the session to include our copy of `NS` function.
    - See suggestion

```suggestion

  ns_local <- NS
  ns_env <- environment(session$ns)
  ns_env$NS <- ns_local

  shiny::moduleServer(id, module, session)
```

Reply:
I think it's better to live with this problem, than to modify the `session` environment manually, this is a poor solution, but I couldn't find an alternative.

Opinions?